modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/MinMuleVersionLoaderDelegatesTestCase.java

```diff
@@ -0,0 +1,346 @@
+import org.mule.runtime.api.exception.MuleException;
+import org.mule.runtime.api.lifecycle.InitialisationException;
+import org.mule.runtime.api.lifecycle.Lifecycle;
+import org.mule.runtime.api.meta.MuleVersion;
+import org.mule.runtime.api.meta.model.declaration.fluent.ConfigurationDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.ConnectionProviderDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.ExtensionDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.ExtensionDeclarer;
+import org.mule.runtime.api.meta.model.declaration.fluent.FunctionDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.OperationDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.ParameterDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.ParameterGroupDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.SourceDeclaration;
+import org.mule.runtime.api.meta.model.parameter.ParameterRole;
+import org.mule.runtime.extension.api.property.SourceClusterSupportModelProperty;
+import org.mule.runtime.extension.api.runtime.config.ConfigurationFactory;
+import org.mule.runtime.extension.internal.loader.DefaultExtensionLoadingContext;
+import org.mule.runtime.module.extension.api.loader.java.type.ExtensionElement;
+import org.mule.runtime.module.extension.internal.loader.delegate.DefaultExtensionModelLoaderDelegate;
+import org.mule.runtime.module.extension.internal.loader.delegate.StereotypeModelLoaderDelegate;
+import org.mule.runtime.module.extension.internal.loader.java.property.ConfigurationFactoryModelProperty;
+import org.mule.runtime.module.extension.internal.loader.parser.ConfigurationModelParser;
+import org.mule.runtime.module.extension.internal.loader.parser.ConnectionProviderModelParser;
+import org.mule.runtime.module.extension.internal.loader.parser.DefaultOutputModelParser;
+import org.mule.runtime.module.extension.internal.loader.parser.ExtensionModelParser;
+import org.mule.runtime.module.extension.internal.loader.parser.FunctionModelParser;
+import org.mule.runtime.module.extension.internal.loader.parser.OperationModelParser;
+import org.mule.runtime.module.extension.internal.loader.parser.OutputModelParser;
+import org.mule.runtime.module.extension.internal.loader.parser.ParameterGroupModelParser;
+import org.mule.runtime.module.extension.internal.loader.parser.ParameterModelParser;
+import org.mule.runtime.module.extension.internal.loader.parser.SourceModelParser;
+import org.mule.runtime.module.extension.internal.loader.parser.java.JavaExtensionModelParser;
+import org.mule.runtime.module.extension.internal.loader.parser.java.JavaExtensionModelParserFactory;
+import org.mule.sdk.api.annotation.Extension;
+import org.mule.sdk.api.annotation.source.SourceClusterSupport;
+
+import java.util.List;
+import java.util.Optional;
+
+public class MinMuleVersionLoaderDelegatesTestCase {
```

Comment:
not getting the point of this test. It's way too hardwired. Most of the code is actually manually building models which then of course will match the assertions.

I would rather have a functional test which takes test extensions and validates that each component has the right versions.

Reply:
This tests make sure that the loader delegates use the parsers to select the min mule version of the models. For this, the parsers used are mocks, this way it does not depend on the java parsers implementation.

I want to know that for all models that can have the minMuleVersion configured, the loaders will ask the parser for the minMuleVersion and configure the declaration with that version.

Reply:
I think that we already have tests for the parsers calculation. This test isolates the parser behavior by creating the mocks and tests the usage of the parsers.

If this test breaks, we know that the issue is not in the parsers logic, and we know that it will affect all SDKs.

What you are proposing is have this test use an already existing implementation of the model parsers?

Reply:
what I'm saying is that tests should focus on the system behavior, resilience performance, etc.. Not in the exact implementation that leads to those things.

If tomorrow we decide to refactor the parsers completely, or getting rid of the loadingDelegate all together, then this test is forced to change alongside it. By doing that, it's not really verifying that no regressions are introduced or compatibility is changed. This is because you're **testing the implementation**. 

On the other side, if you write tests that takes extensions and verify that the versions are correclty resolved, you're testing the behavior. You can completely change how we resolve this, your test is still valid as it's untouched. That's because it's **testing the behaviour.**

Reply:
Got it, I will add examples in the test-extensions and make sure that when the whole model is created the min mule version is the correct one.

What would happen for the case of parameters? we cannot make such test for parameters since there is no way to set a min mule version for them.

Reply:
Ok.

For those cases, I think that the decision is that the operation mule version should be higher instead of the parameters, at least for what I see in the code.